FreeRTOS-Plus/Demo/coreSNTP_Windows_Simulator/DemoTasks/SNTPClientTask.c

```diff
@@ -607,6 +818,371 @@ static void sntpClient_SetTime( const SntpServerInfo_t * pTimeServer,
+                                             ( CK_BYTE_PTR ) pResponseData,
+                                             SNTP_PACKET_BASE_SIZE,
+                                             ( CK_BYTE_PTR ) pResponseData + SNTP_PACKET_SYMMETRIC_KEY_ID_OFFSET + SNTP_PACKET_SYMMETRIC_KEY_ID_LENGTH,
+                                             pkcs11AES_CMAC_SIGNATURE_LENGTH );
+
+            if( result != CKR_OK )
+            {
+                returnStatus = SntpServerNotAuthenticated;
+                LogError( ( "Server cannot be validated from received response: AES-128-CMAC signature in response packet does not match expected." ) );
+            }
+        }
+
+        /* Close the PKCS #11 session as the AES-CMAC operation is completed. */
+        if( result == CKR_OK )
+        {
+            result = functionList->C_CloseSession( pkcs11Session );
+            configASSERT( result == CKR_OK );
+
+            result = functionList->C_Finalize( NULL );
+            configASSERT( result == CKR_OK );
+        }
+    }
+
+    return returnStatus;
+}
+
+
+/*************************************************************************************/
+
+static uint32_t generateRandomNumber()
+{
+    CK_RV pkcs11Status = CKR_OK;
+    CK_FUNCTION_LIST_PTR pFunctionList = NULL;
+    CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+    uint32_t randomNum = 0;
+
+    /* Get list of functions supported by the PKCS #11 port. */
+    pkcs11Status = C_GetFunctionList( &pFunctionList );
+
+    if( pkcs11Status != CKR_OK )
+    {
+        configASSERT( pFunctionList != NULL );
+        LogError( ( "Failed to generate random number. "
+                    "PKCS #11 API, C_GetFunctionList, failed." ) );
+    }
+
+    if( pkcs11Status == CKR_OK )
+    {
+        /* Initialize PKCS #11 module and create a new session. */
+        pkcs11Status = xInitializePkcs11Session( &session );
+
+        if( pkcs11Status != CKR_OK )
+        {
+            configASSERT( session != CK_INVALID_HANDLE );
+
+            LogError( ( "Failed to generate random number. "
+                        "Failed to initialize PKCS #11 session." ) );
+        }
+    }
+
+    if( pkcs11Status == CKR_OK )
+    {
+        if( pFunctionList->C_GenerateRandom( session,
+                                             &randomNum,
+                                             sizeof( randomNum ) ) != CKR_OK )
+        {
+            LogError( ( "Failed to generate random number. "
+                        "PKCS #11 API, C_GenerateRandom, failed to generate random number." ) );
+        }
+    }
+
+    if( pkcs11Status == CKR_OK )
+    {
+        if( pFunctionList->C_CloseSession( session ) != CKR_OK )
+        {
+            LogError( ( " Failed to close PKCS #11 session after generating random number." ) );
+        }
+    }
+
+    return randomNum;
+}
+
+
 /*************************************************************************************/
 
 void initializeSystemClock( void )
```

Comment:
This function does not seem to have a declaration?

Reply:
This function is declared in a common header file, [common_demo_include.h](https://github.com/FreeRTOS/FreeRTOS/blob/main/FreeRTOS-Plus/Demo/coreSNTP_Windows_Simulator/common_demo_include.h#L59-L64), accessed by other files of the demo.